File: solutions/04.enterprise-bot/EnterpriseBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;

namespace EnterpriseBot
{
    /// <summary>
    /// Main entry point and orchestration for bot.
    /// </summary>
    public class EnterpriseBot : IBot
    {
        private readonly BotServices _services;
        private readonly ConversationState _conversationState;
        private readonly UserState _userState;
        private DialogSet _dialogs;

        /// <summary>
        /// Initializes a new instance of the <see cref="EnterpriseBot"/> class.
        /// </summary>
        /// <param name="botServices">Bot services.</param>
        /// <param name="conversationState">Bot conversation state.</param>
        /// <param name="userState">Bot user state.</param>
        public EnterpriseBot(BotServices botServices, ConversationState conversationState, UserState userState)
        {
            _conversationState = conversationState ?? throw new ArgumentNullException(nameof(conversationState));
            _userState = userState ?? throw new ArgumentNullException(nameof(userState));
            _services = botServices ?? throw new ArgumentNullException(nameof(botServices));

            _dialogs = new DialogSet(_conversationState.CreateProperty<DialogState>(nameof(EnterpriseBot)));
            _dialogs.Add(new MainDialog(_services, _conversationState, _userState));
        }

        /// <summary>
        /// Run every turn of the conversation. Handles orchestration of messages.
        /// </summary>
        /// <param name="turnContext">Bot Turn Context.</param>
        /// <param name="cancellationToken">Task CancellationToken.</param>
        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken)
        {
            var dc = await _dialogs.CreateContextAsync(turnContext);
            var result = await dc.ContinueDialogAsync();

            if (result.Status == DialogTurnStatus.Empty)
            {
                await dc.BeginDialogAsync(nameof(MainDialog));
            }
        }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/enterpriseBot.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { ConversationState, TurnContext, UserState } from "botbuilder";
import { DialogSet, DialogState, DialogTurnStatus } from "botbuilder-dialogs";
import { BotServices } from "./botServices";
import { MainDialog } from "./dialogs/main/mainDialog";

/**
 * Main entry point and orchestration for bot.
 */
export class EnterpriseBot {
    private readonly _botServices: BotServices;
    private readonly _conversationState: ConversationState;
    private readonly _userState: UserState;
    private readonly _dialogs: DialogSet;

    /**
     * Constructs the three pieces necessary for this bot to operate.
     * @constructor
     * @param {BotServices} botServices Services included in the bot (LUIS, QnA, Dispatcher, etc...).
     * @param {ConversationState} conversationState Bot conversation state.
     * @param {UserState} userState Bot user state.
     */
    constructor(botServices: BotServices, conversationState: ConversationState, userState: UserState) {
        if (!botServices) { throw new Error(("Missing parameter.  botServices is required")); }
        if (!conversationState) { throw new Error(("Missing parameter.  conversationState is required")); }
        if (!userState) { throw new Error(("Missing parameter.  userState is required")); }

        this._botServices = botServices;
        this._conversationState = conversationState;
        this._userState = userState;

        this._dialogs = new DialogSet(this._conversationState.createProperty<DialogState>("EnterpriseBot"));
        this._dialogs.add(new MainDialog(this._botServices, this._conversationState, this._userState));
    }

    /**
     * Run every turn of the conversation. Handles orchestration of messages.
     * @param {TurnContext} turnContext Turn context from the adapter.
     */
    public async onTurn(turnContext: TurnContext): Promise<void> {
        const dc = await this._dialogs.createContext(turnContext);
        const result = await dc.continueDialog();

        if (result.status === DialogTurnStatus.empty) {
            await dc.beginDialog("MainDialog");
        }
    }
}
